mm2src/lp_ordermatch.rs

```diff
@@ -2153,36 +2156,52 @@ impl From<MakerConnected> for new_protocol::OrdermatchMessage {
     }
 }
 
+fn broadcast_keep_alive_for_pub(ctx: &MmArc, pubkey: &str, orderbook: &Orderbook, p2p_privkey: &Option<H256Json>) {
+    let state = match orderbook.pubkeys_state.get(pubkey) {
+        Some(s) => s,
+        None => return,
+    };
+
+    let mut trie_roots = HashMap::new();
+    let mut topics = HashSet::new();
+    for (alb_pair, root) in state.trie_roots.iter() {
+        if *root == H64::default() && *root == hashed_null_node::<Layout>() {
+            continue;
+        }
+        topics.insert(orderbook_topic_from_ordered_pair(alb_pair));
+        trie_roots.insert(alb_pair.clone(), *root);
+    }
+
+    let message = new_protocol::PubkeyKeepAlive {
+        trie_roots,
+        timestamp: now_ms() / 1000,
+    };
+
+    broadcast_ordermatch_message(ctx, topics, message.into(), p2p_privkey);
+}
+
 pub async fn broadcast_maker_orders_keep_alive_loop(ctx: MmArc) {
-    let my_pubsecp = CryptoCtx::from_ctx(&ctx)
+    let persistent_pubsecp = CryptoCtx::from_ctx(&ctx)
         .expect("CryptoCtx not available")
         .secp256k1_pubkey_hex();
 
     while !ctx.is_stopping() {
         Timer::sleep(MIN_ORDER_KEEP_ALIVE_INTERVAL as f64).await;
         let ordermatch_ctx = OrdermatchContext::from_ctx(&ctx).expect("from_ctx failed");
-        let orderbook = ordermatch_ctx.orderbook.lock();
-        let state = match orderbook.pubkeys_state.get(&my_pubsecp) {
-            Some(s) => s,
-            None => continue,
-        };
+        let my_orders = ordermatch_ctx.my_maker_orders.lock().clone();
+        for (_, order_mutex) in my_orders {
+            let order = order_mutex.lock().await;
```

Comment:
I will recheck this today.

Reply:
As per this, I would like to give it a comprehensive testing on a big maker node. If all works well, we can keep it for a while. In general, it seems that we have to refactor `my_maker_orders` and `orderbook`: *my* orders and *my* pubkey states should be in the same struct that will be covered by a single lock. We can consider this refactoring in future sprints.

Reply:
I'd prefer to just clone the order mutex as it's easy to do and allows us to avoid deadlocks until we refactor `my_maker_orders` and `orderbook`.
```rust
let order = order_mutex.lock().await.clone();
```

Reply:
As we need only p2p_privkey there, I implemented it without cloning the entire order.

Reply:
That's much better, thanks!